### PR TITLE
feat(common): add URLSearchParams to request body

### LINKED ISSUE #36317: HttpClient#post does not serialize native UrlSearchParams
# 🐞 bug report

### Affected Package
The issue is caused by package @angular/common/http

### Is this a regression?

I just started using the package so I'm not sure if it worked for previous versions

### Description

Passing an instance of browser-native `UrlSearchParams` as the second argument to `HttpClient#post` (or as the `body` option for any request with a body) causes an empty JSON object string "{}" to be used as the body, instead of the correctly serialized URL-encoded form string.

## 🔬 Minimal Reproduction
https://angular-issue-repro2-cs1zmj.stackblitz.io

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.0.3
Node: 12.7.0
OS: win32 x64

Angular: 9.0.2
... common, compiler, core
Ivy Workspace: <error>

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.900.3
@angular-devkit/core         9.0.3
@angular-devkit/schematics   9.0.3
@angular/cli                 9.0.3
@schematics/angular          9.0.3
@schematics/update           0.900.3
rxjs                         6.5.3
</code></pre>

**Anything else relevant?**
I've tested this in latest Edge, Firefox, and Chrome.  (IE doesn't support native UrlSearchParams.)

### PR DESCRIPTION
**Pull request by ajitsinghkaler**
URLSearch params are by default supported in the browser but are not supported by angular/http package added support for URLSearchParams

Fixes #36317

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

No support for URL searchParams

Issue Number: 36317

## What is the new behavior?
Support for URL searchParams

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by ajitsinghkaler** [comment]
@gkalpak another size flake

**Comment by ajitsinghkaler** [comment]
@alxhub how to get rid of the error in legacy-unit-tests-saucelabs

File: packages/common/http/src/request.ts [hunk 1]
```diff
@@ -65,6 +65,15 @@ function isFormData(value: any): value is FormData {
   return typeof FormData !== 'undefined' && value instanceof FormData;
 }
 
+/**
+ * Safely assert whether the given value is a URLSearchParams instance.
+ *
+ * In some execution environments URLSearchParams is not defined.
+ */
+function isUrlSearchParams(value: any): value is URLSearchParams {
```
Comment (alfaproject) [inline packages/common/http/src/request.ts:73]:
```suggestion
function isUrlSearchParams(value: unknown): value is URLSearchParams {
```

File: packages/common/http/src/request.ts [hunk 2]
```diff
@@ -65,6 +65,15 @@ function isFormData(value: any): value is FormData {
...
+  return typeof URLSearchParams !== 'undefined' && value instanceof URLSearchParams;
```
Comment (alfaproject) [inline packages/common/http/src/request.ts:74]:
```suggestion
  return value instanceof URLSearchParams;
```

Comment (ajitsinghkaler) [reply packages/common/http/src/request.ts:74]: Why are we not checking for undefined

I will not work on older browsers

Comment (alfaproject) [reply packages/common/http/src/request.ts:74]: You are right, ignore me.

File: packages/common/http/src/request.ts (hunk 1 above)
Comment (petebacondarwin) [reply packages/common/http/src/request.ts:73]: `isBlob` and `isFormData` accept `any`, so I wouldn't be too worried about changing the convention here. There is very little value in this being `unknown` here to be honest.

Instead we could do a subsequent refactoring of all these type guards if we wanted.

**Review by petebacondarwin — APPROVED** [review]
LGTM!

**Comment by petebacondarwin** [comment]
This is actually a duplicate of #37767, which arrived just before this one.
We are going to go with that if the author is able to make the required changes.
Otherwise we'll come back here. In the meantime I am going to close this one to avoid confusion.

Thanks for the contribution @ajitsinghkaler.

**Comment by petebacondarwin** [comment]
Hi @ajitsinghkaler - https://github.com/angular/angular/pull/37767 never got updated, so I am re-opening this one. Are you interested in trying to land it?

**Comment by ajitsinghkaler** [comment]
@petebacondarwin I'm interested can you tell me why ci is failing I think its because tests are not working in legacy browser

Can you tell me how to disable it on legacy browsers? Or if there is anything else

**Comment by petebacondarwin** [comment]
The failing tests are probably flakes. I tried restarting them.
Could you rebase on master and see what happens?

**Comment by ajitsinghkaler** [comment]
Yeah sure

**Comment by ajitsinghkaler** [comment]
@petebacondarwin reased it

**Comment by petebacondarwin** [comment]
Caretaker: I have marked with both presubmit and merge because I am confident that this will sail through G3.

**Pull request closed without merging**